File: App/Services/CoordinateMapper.swift
import CoreGraphics

/// Maps Vision-normalized points (bottom-left origin, y-up) into
/// (a) view coordinates for an aspect-fill portrait preview and
/// (b) PoseKit normalized y-down coordinates for scoring.
struct CoordinateMapper {
    let bufferSize: CGSize   // portrait-oriented buffer dimensions
    let viewSize: CGSize
    let isMirrored: Bool     // true for front camera

    func viewPoint(fromVisionPoint p: CGPoint) -> CGPoint {
        var x = p.x
        if isMirrored { x = 1 - x }
        let yDown = 1 - p.y

        let scale = max(viewSize.width / bufferSize.width,
                        viewSize.height / bufferSize.height)
        let scaled = CGSize(width: bufferSize.width * scale,
                            height: bufferSize.height * scale)
        let offsetX = (viewSize.width - scaled.width) / 2
        let offsetY = (viewSize.height - scaled.height) / 2

        return CGPoint(x: x * scaled.width + offsetX,
                       y: yDown * scaled.height + offsetY)
    }

    func poseKitPoint(fromVisionPoint p: CGPoint) -> SIMD2<Float> {
        let x = isMirrored ? 1 - p.x : p.x
        return SIMD2<Float>(Float(x), Float(1 - p.y))
    }
}

File: android/app/src/main/kotlin/com/oerol/pose/camera/CoordinateMapper.kt
package com.oerol.pose.camera

import com.oerol.posekit.Vec2

/**
 * Maps normalized pose coordinates ([0,1], y-down, already mirrored for the
 * front camera by PoseAnalyzer) into view pixels for an aspect-fill preview —
 * the same center-crop math as the iOS CoordinateMapper, minus the Vision
 * y-flip (ML Kit is already y-down).
 */
class CoordinateMapper(
    private val bufferWidth: Float,   // upright (rotation-applied) buffer dims
    private val bufferHeight: Float,
    private val viewWidth: Float,
    private val viewHeight: Float,
) {
    private val scale = maxOf(viewWidth / bufferWidth, viewHeight / bufferHeight)
    private val offsetX = (viewWidth - bufferWidth * scale) / 2
    private val offsetY = (viewHeight - bufferHeight * scale) / 2

    fun viewX(normalizedX: Float): Float = normalizedX * bufferWidth * scale + offsetX
    fun viewY(normalizedY: Float): Float = normalizedY * bufferHeight * scale + offsetY

    fun viewPoint(p: Vec2): Vec2 = Vec2(viewX(p.x), viewY(p.y))
}
